elastic/elastic-agent
# Clean up / document Agent policy processing

Close https://github.com/elastic/ingest-dev/issues/1913
Document Agent's handling of component spec files and policies: Add documentation to `docs/component-specs.md` and `docs/agent-policy.md` listing the fields that can be used and describing how Agent assembles them into a final component configuration.

Refactor `component.go`:
- Large nested for loops, particularly in `PolicyToComponents`, were broken up into smaller helper functions.
- Several small simplifications to processing, for example:
  * `policy.revision` is now injected into input configurations in `toIntermediate` along with the rest of the global policy edits
  * `inputI.inputType` is now documented as being the canonical input type rather than an alias, and the several later stages of re-canonicalization are removed
  *  `RuntimeSpecs.GetInput` and `getSupportedShipperType` now run `validateRuntimeChecks` as part of their platform verification instead of requiring their callers to do it in a separate stage
- Many fields and helpers have added documentation

Add `use_shipper` support to output configurations: rather than requiring outputs to separately configure each possible supported shipper (of which there is currently only one), and forbidding the output configurations to have any keys that match the name of a shipper (which may change over time), centralize on a single boolean field `use_shipper`. If `use_shipper` is true, then Agent will choose a supported shipper to implement the output. If more than one supported shipper implements that output type, Agent chooses the one that appears first in the input type's `shippers` list in the spec file. (This feature is also simpler to implement, and enables many of the simplifications in the `PolicyToComponents` rewrite.)

This PR should not affect any functional behavior except for policies that enable the shipper (in particular, it should not affect the behavior of any officially released features).


## Checklist


## How to test this PR locally

There were already extensive tests of policy handling, which helped a lot with verifying these changes. The only changes needed were migrating tests to use `use_shipper` instead of the old model based on the shipper name. The biggest semantic change in the migration is that outputs that set `use_shipper: true` will no longer silently fall back to a non-shipper output if the shipper is not supported. It is now considered an error to direct an input at a shipper output it does not support.
